pkg/report/coverageprofile.go

```diff
@@ -0,0 +1,306 @@
+package report
+
+import (
+	"fmt"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/Azure/gocover/pkg/gittool"
+	"golang.org/x/tools/cover"
+)
+
+// DiffCoverage expose the diff coverage statistics
+type DiffCoverage interface {
+	GenerateDiffCoverage() (*Statistics, error)
+}
+
+func NewDiffCoverage(
+	profiles []*cover.Profile,
+	changes []*gittool.Change,
+	excludes []string,
+	comparedBranch string,
+) DiffCoverage {
+	return &diffCoverage{
+		comparedBranch: comparedBranch,
+		excludes:       excludes,
+		profiles:       profiles,
+		changes:        changes,
+	}
+}
+
+var _ DiffCoverage = (*diffCoverage)(nil)
+
+// diffCoverage implements the DiffCoverage interface
+// and generate the diff coverage statistics
+type diffCoverage struct {
+	comparedBranch string            // git diff base branch
+	excludes       []string          // excludes files
+	profiles       []*cover.Profile  // go unit test coverage profiles
+	changes        []*gittool.Change // diff change between compared branch and HEAD commit
+}
+
+func (diff *diffCoverage) GenerateDiffCoverage() (*Statistics, error) {
+	if err := diff.ignore(); err != nil {
+		return nil, err
+	}
+	diff.filter()
+	return diff.percentCovered(), nil
+}
+
+// ignore files that not accountting for diff coverage
+// support standard regular expression
+func (diff *diffCoverage) ignore() error {
+	var filterProfiles []*cover.Profile
+
+	for _, p := range diff.profiles {
+		find := false
+		for _, ignorePattern := range diff.excludes {
+			reg, err := regexp.Compile(ignorePattern)
```

Comment:
please put the regexp.Compile out of the loop.
so it won't be "compiled" for every profile.

Reply:
suggestion: may extract the compile of the excludes into one function,
and make that function return the compiled regex, and the error there.

so the "ignore" function become more clean.

Reply:
Reorgniazed.